jax/interpreters/pxla.py

```diff
@@ -913,112 +936,193 @@ def parallel_callable(fun: lu.WrappedFun,
   else:
     out_tuple = build_out_tuple()
 
-  if backend.platform in ("gpu", "tpu"):
+  if be.platform in ("gpu", "tpu"):
     donated_invars = xla.set_up_aliases(c, xla_args, c.GetShape(out_tuple),
                                         donated_invars, tuple_args)
   built = c.Build(out_tuple)
 
-  if devices is None:
-    if num_global_shards > xb.device_count(backend):
-      msg = ("compiling computation that requires {} logical devices, but only {} XLA "
-             "devices are available (num_replicas={}, num_partitions={})")
-      raise ValueError(msg.format(num_global_shards, xb.device_count(backend),
-                                  num_global_replicas, num_partitions))
-
-    # On a single host, we use the platform's default device assignment to
-    # potentially take advantage of device locality. On multiple hosts, the
-    # default device assignment may interleave different hosts' replicas,
-    # violating pmap's semantics where data is sharded across replicas in
-    # row-major order. Instead, manually create a device assignment that ensures
-    # each host is responsible for a continguous set of replicas.
-    if num_global_shards > num_local_shards:
-      # TODO(skye): use a locality-aware assignment that satisfies the above
-      # constraint.
-      devices = [d for process_index in range(xb.process_count(backend))
-                 for d in xb.local_devices(process_index, backend)]
-    else:
-      devices = xb.get_backend(backend).get_default_device_assignment(
-          num_global_replicas, num_partitions)
-  else:
-    if num_local_shards != len(local_devices):
-      local_devices_str = ", ".join(map(str, local_devices))
-      if num_local_shards == axis_size:
-        raise ValueError(
-            f"Leading axis size of input to pmapped function must equal the "
-            f"number of local devices passed to pmap. Got axis_size="
-            f"{axis_size}, num_local_devices={len(local_devices)}.\n(Local "
-            f"devices available to pmap: {local_devices_str})")
+  return PmapComputation(built,
+                         axis_name,
+                         axis_size,
+                         in_axes,
+                         out_axes,
+                         local_devices,
+                         devices,
+                         be,
+                         tuple_args,
+                         local_num_partitions,
+                         num_partitions,
+                         local_arg_parts,
+                         local_out_parts,
+                         out_parts,
+                         num_local_replicas,
+                         num_local_shards,
+                         num_global_shards,
+                         num_global_replicas,
+                         avals,
+                         sharded_avals,
+                         out_sharded_avals)
```

Comment:
This is a huge struct. Maybe some of the information is redundant. To cut it down in size, maybe some of these values can be computed in terms of others, and don't have to be stored?

(And maybe join these lines instead of having one argument per line, here and elsewhere?)

Reply:
I've done some refactoring. Thoughts?